### PR TITLE
data/env: more workarounds for even older fish shells, provide reasonable defaults

### PR DESCRIPTION
**Pull request by bboozzoo**
Tweak snapd.fish to be compatible with even older releases of fish, going back to 2.7, which was shipped with Ubuntu 18.04.

### CODE REVIEW

File: tests/main/user-session-env/task.yaml [hunk 1]
```diff
@@ -7,32 +7,53 @@ details: |
...
-    for user in test "$TEST_ZSH_USER"; do
+    for user in test "$TEST_ZSH_USER" ; do
         # Dump the environment set up by the user session manager
         if tests.session has-session-systemd-and-dbus; then
             tests.session -u "$user" exec systemctl --user show-environment > "${user}-session-env"
         fi
         tests.session -u "$user" exec env > "${user}-profile-env"
+        # fish allows environment variables to not be exported to subprocesses, make
+        # sure the right variables are exported though
+        tests.session -u "$user" exec sh -c 'exec env' > "${user}-child-env"
```
Comment (mardy) [inline tests/main/user-session-env/task.yaml:46]: This comment about fish is a bit misleading here: maybe rephrase it something like

> Later we will later check that the subprocess has all the variables it needs (not all shells behave in the same way)

And then you have anyway the specific comment about fish below.

File: tests/main/user-session-env/task.yaml [hunk 2]
```diff
@@ -45,13 +66,34 @@ execute: |
         # Profile should also be correctly set up
         case "$user:$SPREAD_SYSTEM" in
             test-zsh:ubuntu-*|test-zsh:debian-*)
-                # Due to https://bugs.launchpad.net/ubuntu/+source/zsh/+bug/1640514
+                # zsh due to https://bugs.launchpad.net/ubuntu/+source/zsh/+bug/1640514
```
Comment (mardy) [inline tests/main/user-session-env/task.yaml:69]: Is this a search&replace issue? The old comment looks OK to me.

**Review by mardy — APPROVED** [review]
Minor nitpicks, but LGTM!

**Review by stolowski — APPROVED** [review]
LGTM

File: data/env/snapd.fish.in [hunk 3]
```diff
@@ -8,7 +8,14 @@ end
...
 set -ul snap_xdg_path /var/lib/snapd/desktop
-set --path --export XDG_DATA_DIRS $XDG_DATA_DIRS
-if not contains $snap_xdg_path $XDG_DATA_DIRS
-    set XDG_DATA_DIRS $XDG_DATA_DIRS $snap_xdg_path
+# note, snapd supports distros going back as far as Ubuntu 16.04, what means we
+# must make sure that this is compatible with old fish shells which do not have
+# fish_add_path or set --path
+set --export XDG_DATA_DIRS $XDG_DATA_DIRS
+if not contains $snap_xdg_path (string split : "$XDG_DATA_DIRS")
+    if test (string length "$XDG_DATA_DIRS") -gt 0
```
Comment (faho) [inline data/env/snapd.fish.in:16]: This entire block can just be

```fish
set XDG_DATA_DIRS (string join : -- $XDG_DATA_DIRS $snap_xdg_path)
```

If $XDG_DATA_DIRS is empty, it will be elided and only $snap_xdg_path remains.

Or alternatively replace the if-condition:

```fish
if set -q XDG_DATA_DIRS[1]
```

File: data/env/snapd.fish.in [hunk 4]
```diff
@@ -8,7 +8,14 @@ end
 # looked for in XDG_DATA_DIRS; make sure it includes the relevant directory for
 # snappy applications' desktop files.
 set -ul snap_xdg_path /var/lib/snapd/desktop
-set --path --export XDG_DATA_DIRS $XDG_DATA_DIRS
-if not contains $snap_xdg_path $XDG_DATA_DIRS
-    set XDG_DATA_DIRS $XDG_DATA_DIRS $snap_xdg_path
+# note, snapd supports distros going back as far as Ubuntu 16.04, what means we
+# must make sure that this is compatible with old fish shells which do not have
+# fish_add_path or set --path
+set --export XDG_DATA_DIRS $XDG_DATA_DIRS
```
Comment (faho) [inline data/env/snapd.fish.in:14]: In fish, "exported" is not a scope. You probably want to explicitly request a global variable:

```fish
set --global --export XDG_DATA_DIRS $XDG_DATA_DIRS
```

Comment (faho) [reply data/env/snapd.fish.in:14]: (this isn't necessary for $PATH because that's guaranteed to be global)

**Comment by bboozzoo** [comment]
@faho thanks for the review, I'll push an update in a bit.

**Comment by bboozzoo** [comment]
Marked as blocked as we may need to address https://bugs.launchpad.net/ubuntu/+source/snapd/+bug/1960702 too.
